### PR TITLE
[ROCm] enable unit tests and other changes

### PR DESCRIPTION
**Pull request by iotamudelta**
This PR for the ROCm target does the following:
* enable some unit tests on ROCm
* fix a missing static_cast that breaks BatchNorm call on ROCm
* fix BatchNorm to work on ROCm w/ ROCm warp sizes etc
* improve the pyhipify script by introducing kernel scope to some transpilations and other improvements
* fix a linking issue on ROCm
* for more unit test sets: mark currently broken tests broken (to be fixed)
* enable THINLTO (phase one) to parallelize linking
* address the first failing of the elementwise kernel by removing non-working ROCm specialization

### CODE REVIEW

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -30,6 +30,7 @@ cmake --version
 pip install -r requirements.txt || true
 
 if [[ "$BUILD_ENVIRONMENT" == *rocm* ]]; then
+  # This is necessary in order to cross compile (or else we'll have missing GPU device).
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: I don't think this comment should have been duplicated here?

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 2]
```diff
@@ -28,11 +28,7 @@ namespace at { namespace native {
 
 template<int nt, int vt, typename func_t>
 __launch_bounds__(nt, 4)
-#ifdef __HIP_PLATFORM_HCC__
-__global__ void elementwise_kernel(int N, const func_t& f) {
-#else
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/Loops.cuh]: Nice!

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -341,13 +341,24 @@ endif()
 # ---[ Caffe2 HIP sources.
 if(USE_ROCM)
   # Call again since Caffe2_HIP_INCLUDES is extended with ATen include dirs.
-  IF(BUILD_ATEN)
-     HIP_INCLUDE_DIRECTORIES(${Caffe2_HIP_INCLUDES})
-  ENDIF()
+  if(BUILD_ATEN)
+    # Get Compile Definitions from the directory (FindHIP.CMake bug)
```
Comment (ezyang) [inline caffe2/CMakeLists.txt]: nit: FindHIP.cmake

File: test/test_autograd.py [hunk 4]
```diff
@@ -975,6 +975,7 @@ def test_no_requires_grad_inplace(self):
         with self.assertRaises(RuntimeError):
             b.add_(5)
 
+    @unittest.skipIf(TEST_WITH_ROCM, "test doesn't currently work on the ROCm stack")
```
Comment (ezyang) [inline test/test_autograd.py]: If you defined `skipIfRocm`, why don't you use it?

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 5]
```diff
@@ -426,21 +427,27 @@ def find_kernel_bounds(string):
     return output_string
 
 
-def find_parenthesis_end(input_string, start):
+def find_closure_group(input_string, start, group):
+    """Generalization for finding a balancing closure group
+
+    e.g. if group = ["(", ")"], then finds the first balanced parantheses.
+         if group = ["{", "}"], then finds the first balanced bracket.
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: OK, but what does it return in this case?

Here, an example is worth a million words:

```
>>> find_closure_group("foo(bar)", 0, ["(", ")"])
????
```

File: tools/amd_build/pyHIPIFY/hipify-python.py [hunk 6]
```diff
@@ -487,6 +504,34 @@ def replace_math_functions(input_string):
     return output_string
 
 
+def hip_header_magic(input_string):
+    """If the file makes kernel builtin calls and does not include the cuda_runtime.h header,
+    then automatically add an #include to match the "magic" includes provided by NVCC.
+    TODO:
+        Update logic to ignore cases where the cuda_runtime.h is included by another file.
```
Comment (ezyang) [inline tools/amd_build/pyHIPIFY/hipify-python.py]: Is this really necessary? It should be harmless to include a header multiple times?

File: tools/build_pytorch_libs.sh [hunk 7]
```diff
@@ -97,7 +97,11 @@ if [[ $(uname) == 'Darwin' ]]; then
     LDFLAGS="$LDFLAGS -Wl,-rpath,@loader_path"
     LD_POSTFIX=".dylib"
 else
-    LDFLAGS="$LDFLAGS -Wl,-rpath,\$ORIGIN"
+    if [[ $USE_ROCM -eq 1 ]]; then
+        LDFLAGS="$LDFLAGS -Wl,-rpath,\\\\\\\$ORIGIN"
```
Comment (ezyang) [inline tools/build_pytorch_libs.sh]: uh... what?!

**Review by ezyang — APPROVED** [review]
Some nits but I don't see any show stoppers.

File: test/test_autograd.py (hunk 4 above)
Comment (iotamudelta) [reply test/test_autograd.py]: @jithunnair-amd comment/fix, thanks!

File: tools/build_pytorch_libs.sh (hunk 7 above)
Comment (iotamudelta) [reply tools/build_pytorch_libs.sh]: @mwootton please comment here, thanks!

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 6 above)
Comment (iotamudelta) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: it is harmless, I think this is strictly in the "beautification" realm.

File: tools/amd_build/pyHIPIFY/hipify-python.py (hunk 5 above)
Comment (iotamudelta) [reply tools/amd_build/pyHIPIFY/hipify-python.py]: @Jorghi12 can you comment?

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (iotamudelta) [reply .jenkins/pytorch/build.sh]: probably happened in a merge. do you want me to update for this PR or just fix for a future one?

File: test/test_autograd.py (hunk 4 above)
Comment (jithunnair-amd) [reply test/test_autograd.py]: We had been able to make do without defining it until I had to define it for test_gesv and some other tests. I will make another PR to clean up all uses of TEST_WITH_ROCM and replace it with the skipIfRocm annotation. I didn't want to clutter up this PR. Let me know if that doesn't work for you.
